hooks/process_items.py

```diff
@@ -0,0 +1,50 @@
+# Copyright (c) 2017 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import sgtk
+
+HookBaseClass = sgtk.get_hook_baseclass()
+
+class ProcessItems(HookBaseClass):
+    """
+    This hook defines methods that operates on top-level publish items during
+    the publishing process.
+    """
+
+    def post_finalize(self, items):
```

Comment:
Thinking about items vs tasks and the fact that we are basically passing a tree of items here (the top-level bit).

Don't you think it would be useful/easier for TD if we passed in a data structure which resembled the tree in the UI? With items and tasks (w settings) - for tasks, basically use the `Task` class - this would even allow for people to add their methods to a publish plugin hook (e.g. `post_finalize()` and then go `task.plugin.post_finalize()` - something like that.

Maybe worth a discussion? It would be really great if this was conceptually really really clear and easy to explain!

If we are passing in a top level object of a tree, maybe this could contain the state for the entire 'session', e.g. info about if the publish failed for example.

Reply:
Thought I'd chip in here as I'm keen to see a post-publish type hook for publish2! 
In regards to your specific question above, I think being able to process the item tree would be great, but it would also be nice to have an items_as_list() function to have a simple list to iterate over. 
Is there a simple method to iterate over the item tree that I've not discovered yet? 

My use case for the post-publish hook is this:

For all heavy tasks; file copies, version creation(qt rendering), etc I create json job descriptions that I compile in to a single 'job-package' that gets submitted to deadline once the publish has completed. 

With the old publisher this was straight forward to achieve a there was a post-publish hook that I could use to submit the job-package to deadline. 

With publish2 this is no longer possible, as I don't have a post-publish hook. 
Thanks
p.